onnxruntime/core/framework/session_state.cc

```diff
@@ -156,20 +186,138 @@ static int64_t CalculateMemoryPatternsKey(const std::vector<std::reference_wrapp
+    if (!shape || shape->dim_size() != static_cast<int>(it->second.NumDimensions()))
+      return Status(ONNXRUNTIME, FAIL, "Graph input " + input->Name() +
+                                           "'s shape is not present or its shape doesn't match feed's shape."
+                                           "Unable to resolve the value for dynamic shape");
+    for (int k = 0; k < shape->dim_size(); ++k) {
+      if (shape->dim()[k].has_dim_param()) {
+        out.insert({shape->dim()[k].dim_param(), it->second.GetDims()[k]});
+      }
+    }
+  }
+  return Status::OK();
+}
+}  // namespace
+
+Status SessionState::GeneratePatternGroupCache(const std::vector<std::reference_wrapper<const TensorShape>>& input_shape,
+                                               const std::vector<int>& feed_mlvalue_idxs,
+                                               MemoryPatternGroup* output) const {
+  std::map<std::string, TensorShape> feeds;
+  for (size_t i = 0; i < feed_mlvalue_idxs.size(); ++i) {
+    std::string name;
+    ORT_RETURN_IF_ERROR(this->ort_value_name_idx_map_.GetName(feed_mlvalue_idxs[i], name));
+    feeds.insert({name, input_shape[i]});
+  }
+  std::unordered_map<std::string, int64_t> map;
+  ORT_RETURN_IF_ERROR(ResolveDimParams(*graph_viewer_, feeds, map));
+  auto* exe_plan = GetExecutionPlan();
+  ORT_ENFORCE(exe_plan);
+  OrtValuePatternPlanner mem_planner(*exe_plan);
+  auto& node_index_info = GetNodeIndexInfo();
+  for (auto& node_plan : exe_plan->execution_plan) {
+    int node_index = node_index_info.GetNodeOffset(node_plan.node_index);
+    auto* node = graph_viewer_->GetNode(node_plan.node_index);
+    int output_start = node_index + static_cast<int>(node->InputDefs().size()) + static_cast<int>(node->ImplicitInputDefs().size());
+    //allocate output
+    for (int i = 0; i < static_cast<int>(node->OutputDefs().size()); ++i) {
+      const auto ml_value_idx = node_index_info.GetMLValueIndex(output_start + i);
+      if (ml_value_idx == NodeIndexInfo::kInvalidEntry)
+        continue;
+      const auto* ml_type = exe_plan->allocation_plan[ml_value_idx].value_type;
+      if (!ml_type->IsTensorType())
+        continue;
+      const auto* ml_data_type = static_cast<const TensorTypeBase*>(ml_type)->GetElementType();
+      if (exe_plan->allocation_plan[ml_value_idx].alloc_kind == AllocKind::kAllocate &&
+          ml_data_type != DataTypeImpl::GetType<std::string>()) {
+        //calculate size
+        auto* arg = node->OutputDefs()[i];
+        if (!arg->Shape())
+          continue;
+        size_t size = 0;
+        SafeInt<size_t> len = 1;
+        for (auto& dim : arg->Shape()->dim()) {
+          if (dim.has_dim_param()) {
+            auto it = map.find(dim.dim_param());
+            if (it == map.end()) {
+              return Status(ONNXRUNTIME, FAIL, "Unknown shape found in memory pattern compute");
+            }
+            len *= it->second;
+          } else {
+            len *= dim.dim_value();
+          }
```

Comment:
It's valid for a dim to have neither dim_param nor dim_value set, so you can't assume dim_value is set without calling has_dim_value() first.